### ISSUE TITLE
Remix builds slow with projects containing a high number of routes

### ISSUE THREAD

**Issue by daniloteodoro** [remix-run/remix#6630]
### What version of Remix are you using?

1.17.1

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

Generate a vanilla Remix app using some basic template like `indie-stack`
Add 300 routes. In my case I added them on the `remix.config.js` file:
```JS
const remixRoutes = require('./data/remixRoutes.json');

module.exports = {

  routes: (defineRoutes) => {
    return defineRoutes((route) => {
      remixRoutes.forEach(([routePath, filePath]) => {
        route(routePath, filePath, {id: routePath});
      });
    });
  },

```
with `remixRoutes.json` having the following structure:
```JSON
[
  ["/path1", "routes/helloworld.tsx"],
  ["/path2", "routes/helloworld.tsx"],
  ["/path3", "routes/helloworld.tsx"],
...
  ["/path300", "routes/helloworld.tsx"]
]
```

Run the app with `npm run dev`

### Expected Behavior

Build times, especially after a change in the source-code happens in under 2 seconds

### Actual Behavior

Build times are 30+ seconds. Rebuild time (hmr) is similar 😭

**Comment by daniloteodoro** [comment]
@jacob-ebey

**Comment by pcattori** [comment]
Confirmed fixed by #6629

## Setup

```js
// remix.config.js
module.exports = {
  routes: (defineRoutes) => {
    return defineRoutes((route) => {
      let i = 1;
      while (i <= 300) {
        let routePath = `/route_${i}`;
        route(routePath, "routes/helloworld.tsx", { id: routePath });
        i += 1;
      }
    });
  },
};
```

```tsx
// app/routes/helloworld.tsx
export default function Component() {
  return <h1>heyooo</h1>;
}
```

## Before

<img width="699" alt="Screenshot 2023-06-17 at 2 57 47 PM" src="https://github.com/remix-run/remix/assets/1477317/962adbc8-dea8-4701-9c73-5b2afe0af7d1">

## After

<img width="694" alt="Screenshot 2023-06-17 at 2 57 52 PM" src="https://github.com/remix-run/remix/assets/1477317/c69cb0a3-cb1c-42bb-901d-29bb26f1a00c">
